File: audiobridge/android/app/src/main/java/app/audiobridge/net/ClockSync.kt
package app.audiobridge.net

import android.os.SystemClock
import org.json.JSONObject
import kotlin.math.abs
import kotlin.math.roundToLong

/**
 * Maps the source monotonic clock to this device's monotonic clock.
 *
 * Low-RTT samples establish the offset. A bounded drift estimate then keeps the
 * speakers aligned during long sessions instead of assuming that every device's
 * oscillator runs at exactly the same speed.
 */
class ClockSync(private val send: (JSONObject) -> Unit) {
    @Volatile private var ready = false
    @Volatile private var anchorLocalUs = 0L
    @Volatile private var anchorOffsetUs = 0.0
    @Volatile private var driftPpm = 0.0
    private var bestRttUs = Long.MAX_VALUE
    private var driftSampleLocalUs = 0L
    private var driftSampleOffsetUs = 0.0

    val offsetUs: Long?
        get() {
            if (!ready) return null
            val now = nowUs()
            return (anchorOffsetUs +
                (now - anchorLocalUs) * driftPpm / 1_000_000.0).roundToLong()
        }

    fun nowUs(): Long = SystemClock.elapsedRealtimeNanos() / 1000

    fun tick() {
        synchronized(this) {
            if (bestRttUs != Long.MAX_VALUE) {
                bestRttUs = (bestRttUs + 2_000).coerceAtMost(200_000)
            }
        }
        send(JSONObject().put("type", "clk").put("t0", nowUs()))
    }

    /** Handles {"type":"clk","t0":localSend,"t1":sourceClock} replies. */
    @Synchronized
    fun onReply(t0: Long, t1: Long) {
        val t2 = nowUs()
        val rtt = t2 - t0
        if (rtt < 0 || rtt > 2_000_000) return
        if (bestRttUs != Long.MAX_VALUE && rtt > bestRttUs + 5_000) return
        bestRttUs = minOf(bestRttUs, rtt)

        val measuredOffset = (t0 + t2) / 2.0 - t1
        if (!ready) {
            anchorLocalUs = t2
            anchorOffsetUs = measuredOffset
            driftSampleLocalUs = t2
            driftSampleOffsetUs = measuredOffset
            ready = true
            return
        }

        val predicted = anchorOffsetUs +
            (t2 - anchorLocalUs) * driftPpm / 1_000_000.0
        val residual = measuredOffset - predicted
        if (abs(residual) > 200_000) {
            // Sleep/resume can invalidate the old relationship. Re-lock quickly.
            anchorLocalUs = t2
            anchorOffsetUs = measuredOffset
            driftPpm = 0.0
            driftSampleLocalUs = t2
            driftSampleOffsetUs = measuredOffset
            return
        }

        val sampleSpan = t2 - driftSampleLocalUs
        if (sampleSpan >= 5_000_000) {
            val measuredPpm = ((measuredOffset - driftSampleOffsetUs) *
                1_000_000.0 / sampleSpan).coerceIn(-250.0, 250.0)
            driftPpm = driftPpm * 0.9 + measuredPpm * 0.1
            driftSampleLocalUs = t2
            driftSampleOffsetUs = measuredOffset
        }
        anchorLocalUs = t2
        anchorOffsetUs = predicted + residual * 0.2
    }

    @Synchronized
    fun reset() {
        ready = false
        anchorLocalUs = 0
        anchorOffsetUs = 0.0
        driftPpm = 0.0
        bestRttUs = Long.MAX_VALUE
        driftSampleLocalUs = 0
        driftSampleOffsetUs = 0.0
    }
}

File: audiobridge/windows/ClockSync.cs
namespace AudioBridge.Win;

/// <summary>
/// Continuously maps the source device monotonic clock to this PC's monotonic clock.
/// Low-RTT samples establish the offset; a bounded drift estimate prevents speakers
/// from slowly separating during long playback sessions.
/// </summary>
public sealed class ClockSync
{
    private readonly object _gate = new();
    private long _bestRttUs = long.MaxValue;
    private long _anchorLocalUs;
    private double _anchorOffsetUs;
    private double _driftPpm;
    private bool _ready;
    private long _lastSampleLocalUs;
    private double _lastSampleOffsetUs;

    public long? OffsetUs
    {
        get
        {
            lock (_gate)
            {
                if (!_ready) return null;
                long now = Clock.Us;
                double predicted = _anchorOffsetUs +
                    (now - _anchorLocalUs) * _driftPpm / 1_000_000d;
                return (long)Math.Round(predicted);
            }
        }
    }

    public void AgeBestSample()
    {
        lock (_gate)
        {
            if (_bestRttUs != long.MaxValue)
                _bestRttUs = Math.Min(_bestRttUs + 2_000, 200_000);
        }
    }

    public void OnReply(long t0, long sourceT1)
    {
        long t2 = Clock.Us;
        long rtt = t2 - t0;
        if (rtt < 0 || rtt > 2_000_000) return;
        double measuredOffset = (t0 + t2) / 2d - sourceT1;

        lock (_gate)
        {
            if (_bestRttUs != long.MaxValue && rtt > _bestRttUs + 5_000) return;
            _bestRttUs = Math.Min(_bestRttUs, rtt);

            if (!_ready)
            {
                _anchorLocalUs = t2;
                _anchorOffsetUs = measuredOffset;
                _lastSampleLocalUs = t2;
                _lastSampleOffsetUs = measuredOffset;
                _ready = true;
                return;
            }

            double predicted = _anchorOffsetUs +
                (t2 - _anchorLocalUs) * _driftPpm / 1_000_000d;
            double residual = measuredOffset - predicted;

            // Resume/sleep can move the relationship abruptly. Re-lock instead of
            // spending minutes slewing toward a stale offset.
            if (Math.Abs(residual) > 200_000)
            {
                _anchorLocalUs = t2;
                _anchorOffsetUs = measuredOffset;
                _driftPpm = 0;
                _lastSampleLocalUs = t2;
                _lastSampleOffsetUs = measuredOffset;
            }
            else
            {
                long sampleSpan = t2 - _lastSampleLocalUs;
                if (sampleSpan >= 5_000_000)
                {
                    double measuredPpm = (measuredOffset - _lastSampleOffsetUs) *
                        1_000_000d / sampleSpan;
                    measuredPpm = Math.Clamp(measuredPpm, -250d, 250d);
                    _driftPpm = _driftPpm * 0.9d + measuredPpm * 0.1d;
                    _lastSampleLocalUs = t2;
                    _lastSampleOffsetUs = measuredOffset;
                }
                _anchorLocalUs = t2;
                _anchorOffsetUs = predicted + residual * 0.2d;
            }
        }
    }

    public void Reset()
    {
        lock (_gate)
        {
            _bestRttUs = long.MaxValue;
            _anchorLocalUs = 0;
            _anchorOffsetUs = 0;
            _driftPpm = 0;
            _ready = false;
            _lastSampleLocalUs = 0;
            _lastSampleOffsetUs = 0;
        }
    }
}
